# Udemy 【한글자막】머신러닝의 모든 것 with Python, R


In [ ]:
from google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive


# Multiple Linear Regression(다중 선형 회귀)



> - 강의 56 - 다중 선형 회귀 1단계
- 강의 57 - 다중 선형 회귀 2단계
- 강의 58 - 다중 선형 회귀 3단계
- 강의 59 - 다중 선형 회귀 4단계

## Importing the libraries

In [ ]:
import numpy as np
import matplotlib.pyplot as plt
import pandas as pd

## Importing the dataset

In [ ]:
pd.read_csv('/content/drive/MyDrive/UDEMY/Machine Learning A-Z (Codes and Datasets)/Part 2 - Regression(회귀)/Section 5 - Multiple Linear Regression(다중 선형 회귀)/50_Startups.csv')
dataset =pd.read_csv('/content/drive/MyDrive/UDEMY/Machine Learning A-Z (Codes and Datasets)/Part 2 - Regression(회귀)/Section 5 - Multiple Linear Regression(다중 선형 회귀)/50_Startups.csv')
dataset.head()

R&D Spend  Administration  Marketing Spend       State     Profit
0  165349.20       136897.80        471784.10    New York  192261.83
1  162597.70       151377.59        443898.53  California  191792.06
2  153441.51       101145.55        407934.54     Florida  191050.39
3  144372.41       118671.85        383199.62    New York  182901.99
4  142107.34        91391.77        366168.42     Florida  166187.94

In [ ]:
dataset.info()
print('=============================')
dataset.isna().sum()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50 entries, 0 to 49
Data columns (total 5 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   R&D Spend        50 non-null     float64
 1   Administration   50 non-null     float64
 2   Marketing Spend  50 non-null     float64
 3   State            50 non-null     object 
 4   Profit           50 non-null     float64
dtypes: float64(4), object(1)
memory usage: 2.1+ KB


R&D Spend          0
Administration     0
Marketing Spend    0
State              0
Profit             0
dtype: int64

In [ ]:
dataset['State'].value_counts()

# State 컬럼 문자열
# 서로 우위가 없는 범주형 변수 
# 원-핫 인코딩을 통해 열 추가해야함.

New York      17
California    17
Florida       16
Name: State, dtype: int64

In [ ]:
# y값
dataset['Profit'].values

array([192261.83, 191792.06, 191050.39, 182901.99, 166187.94, 156991.12,
       156122.51, 155752.6 , 152211.77, 149759.96, 146121.95, 144259.4 ,
       141585.52, 134307.35, 132602.65, 129917.04, 126992.93, 125370.37,
       124266.9 , 122776.86, 118474.03, 111313.02, 110352.25, 108733.99,
       108552.04, 107404.34, 105733.54, 105008.31, 103282.38, 101004.64,
        99937.59,  97483.56,  97427.84,  96778.92,  96712.8 ,  96479.51,
        90708.19,  89949.14,  81229.06,  81005.76,  78239.91,  77798.83,
        71498.49,  69758.98,  65200.33,  64926.08,  49490.75,  42559.73,
        35673.41,  14681.4 ])

In [ ]:
dataset.iloc[:,:-1]
# [행, 열(처음부터 Profit전까지)]

In [ ]:
X = dataset.iloc[:, :-1].values

y = dataset['Profit'].values
# y = dataset.iloc[:, -1].values

In [ ]:
print(X)

[[165349.2 136897.8 471784.1 'New York']
 [162597.7 151377.59 443898.53 'California']
 [153441.51 101145.55 407934.54 'Florida']
 [144372.41 118671.85 383199.62 'New York']
 [142107.34 91391.77 366168.42 'Florida']
 [131876.9 99814.71 362861.36 'New York']
 [134615.46 147198.87 127716.82 'California']
 [130298.13 145530.06 323876.68 'Florida']
 [120542.52 148718.95 311613.29 'New York']
 [123334.88 108679.17 304981.62 'California']
 [101913.08 110594.11 229160.95 'Florida']
 [100671.96 91790.61 249744.55 'California']
 [93863.75 127320.38 249839.44 'Florida']
 [91992.39 135495.07 252664.93 'California']
 [119943.24 156547.42 256512.92 'Florida']
 [114523.61 122616.84 261776.23 'New York']
 [78013.11 121597.55 264346.06 'California']
 [94657.16 145077.58 282574.31 'New York']
 [91749.16 114175.79 294919.57 'Florida']
 [86419.7 153514.11 0.0 'New York']
 [76253.86 113867.3 298664.47 'California']
 [78389.47 153773.43 299737.29 'New York']
 [73994.56 122782.75 303319.26 'Florida']
 [67532

## Encoding categorical data

In [ ]:
# ColumnTransformer('transformer='로 어떤 열의 인덱스에변환을 할것인지 , 'remainder='는 변환이 적용되지 않고 유지되길원하는 값)
# transformers=[('변환의 종류', 인코딩을 실행할 클래스의 이름, onehot인코딩 적용할 열의 인덱스[3 = State])]

In [ ]:
from sklearn.compose import ColumnTransformer
from sklearn.preprocessing import OneHotEncoder
ct = ColumnTransformer(transformers=[('encoder', OneHotEncoder(), [3])], remainder='passthrough')
X = np.array(ct.fit_transform(X))

In [ ]:
# 원핫인코딩 후 
print(X)

[[0.0 0.0 1.0 165349.2 136897.8 471784.1]
 [1.0 0.0 0.0 162597.7 151377.59 443898.53]
 [0.0 1.0 0.0 153441.51 101145.55 407934.54]
 [0.0 0.0 1.0 144372.41 118671.85 383199.62]
 [0.0 1.0 0.0 142107.34 91391.77 366168.42]
 [0.0 0.0 1.0 131876.9 99814.71 362861.36]
 [1.0 0.0 0.0 134615.46 147198.87 127716.82]
 [0.0 1.0 0.0 130298.13 145530.06 323876.68]
 [0.0 0.0 1.0 120542.52 148718.95 311613.29]
 [1.0 0.0 0.0 123334.88 108679.17 304981.62]
 [0.0 1.0 0.0 101913.08 110594.11 229160.95]
 [1.0 0.0 0.0 100671.96 91790.61 249744.55]
 [0.0 1.0 0.0 93863.75 127320.38 249839.44]
 [1.0 0.0 0.0 91992.39 135495.07 252664.93]
 [0.0 1.0 0.0 119943.24 156547.42 256512.92]
 [0.0 0.0 1.0 114523.61 122616.84 261776.23]
 [1.0 0.0 0.0 78013.11 121597.55 264346.06]
 [0.0 0.0 1.0 94657.16 145077.58 282574.31]
 [0.0 1.0 0.0 91749.16 114175.79 294919.57]
 [0.0 0.0 1.0 86419.7 153514.11 0.0]
 [1.0 0.0 0.0 76253.86 113867.3 298664.47]
 [0.0 0.0 1.0 78389.47 153773.43 299737.29]
 [0.0 1.0 0.0 73994.56 122782.75 3

## Splitting the dataset into the Training set and Test set

In [ ]:
from sklearn.model_selection import train_test_split

# 훈련, 테스트 데이터
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size = 0.2, random_state = 0)

## Training the Multiple Linear Regression model on the Training set

In [ ]:
from sklearn.linear_model import LinearRegression
regressor = LinearRegression()
regressor.fit(X_train, y_train)

LinearRegression()

## Predicting the Test set results
- 첫번째 백터 : 훈련 세트에 있는  10개의 실제 이익 백터
    - 50개 중에서 20% 차지

- 두번째 백터 : 같은 훈련 세트에서 10개의 예측 이익이라 각 단계를 비교
- 훈련 세트 안의 예측 이익이 실제 이익과 같은지 확인.

In [ ]:
# 예측
# 예측 백터
# 이익백터 = 회귀.예측()
y_pred = regressor.predict(X_test)
# 소수점 두자리만 표시
np.set_printoptions(precision=2)
# 실제이익백터와 예측이익백터 구래를 나란히 
# 넘파이 배열 합치기 concatenate
# y_pred백터를 len y_pred 열이 있는 배열로 재배열 하고싶다.
# 축 : 0은 세로 / 1은 가로 
print(np.concatenate((y_pred.reshape(len(y_pred),1), y_test.reshape(len(y_test),1)),1))

[[103015.2  103282.38]
 [132582.28 144259.4 ]
 [132447.74 146121.95]
 [ 71976.1   77798.83]
 [178537.48 191050.39]
 [116161.24 105008.31]
 [ 67851.69  81229.06]
 [ 98791.73  97483.56]
 [113969.44 110352.25]
 [167921.07 166187.94]]


# BONUS

## Q1
Making a single prediction (for example the profit of a startup with R&D Spend = 160000, Administration Spend = 130000, Marketing Spend = 300000 and State = 'California')
- 질문 1: 다중 선형 회귀 모델을 사용하여 단일 예측<br>(예: R&D 지출 = 160000, 관리 지출 = 130000, 마케팅 지출 = 300000 및 주 = 캘리포니아인 스타트업의 이익)을 만드는 방법은 무엇입니까?

In [ ]:
# Q1
print(regressor.predict([[1, 0, 0, 160000, 130000, 300000]]))

[181566.92]


Therefore, our model predicts that the profit of a Californian startup which spent 160000 in R&D, 130000 in Administration and 300000 in Marketing is $ 181566,92.

**Important note 1:** Notice that the values of the features were all input in a double pair of square brackets. That's because the "predict" method always expects a 2D array as the format of its inputs. And putting our values into a double pair of square brackets makes the input exactly a 2D array. Simply put:

따라서 우리 모델은 연구 개발에 160000, 관리에 130000, 마케팅에 300000을 투자한 캘리포니아 스타트업의 수익이 181566,92달러라고 예측합니다.

참고 1: 형상의 값은 모두 두 쌍의 대괄호로 입력되었다.<br> "예측" 방법은 항상 2D 배열을 입력 형식으로 예상하기 때문입니다.<br> 
그리고 우리의 값을 두 쌍의 사각 괄호 안에 넣으면 정확히 2D 배열이 됩니다.<br>간단히 말하면:

$1, 0, 0, 160000, 130000, 300000 \rightarrow \textrm{scalars}$

$[1, 0, 0, 160000, 130000, 300000] \rightarrow \textrm{1D array}$

$[[1, 0, 0, 160000, 130000, 300000]] \rightarrow \textrm{2D array}$

**Important note 2:** Notice also that the "California" state was not input as a string in the last column but as "1, 0, 0" in the first three columns. That's because of course the predict method expects the one-hot-encoded values of the state, and as we see in the second row of the matrix of features X, "California" was encoded as "1, 0, 0". And be careful to include these values in the first three columns, not the last three ones, because the dummy variables are always created in the first columns.

중요한 참고 2: "캘리포니아" 상태는 마지막 열에 입력되지 않았지만<br> 첫 번째 열에 0"이 입력되지 않았습니다.<br>물론 예측 방법 중 하나 이상의 원핫코딩 값을 예측하기 때문에 <br>"캘리포니아"의 두 번째 행의 매트릭스(C)의 행렬)의 두 번째 행으로 인코딩되었다.<br>그리고 첫 번째 열에서 이러한 값을 포함할 수 있으므로, 
더미 변수가 첫 번째 열에 생성되므로, 더미 변수가 아닙니다.

## Q2
Getting the final linear regression equation with the values of the coefficients

- 질문 2: 계수의 최종 값을 사용하여 최종 회귀 방정식<br> y = b0 + b1 x1 + b2 x2 + ...를 어떻게 얻습니까?

In [ ]:
print(regressor.coef_)
print(regressor.intercept_)

[ 8.66e+01 -8.73e+02  7.86e+02  7.73e-01  3.29e-02  3.66e-02]
42467.52924853204


Therefore, the equation of our multiple linear regression model is:

따라서 다중 선형 회귀 모델의 방정식은 다음과 같습니다.

$$\textrm{Profit} = 86.6 \times \textrm{Dummy State 1} - 873 \times \textrm{Dummy State 2} + 786 \times \textrm{Dummy State 3} + 0.773 \times \textrm{R&D Spend} + 0.0329 \times \textrm{Administration} + 0.0366 \times \textrm{Marketing Spend} + 42467.53$$

**Important Note:** To get these coefficients we called the "coef_" and "intercept_" attributes from our regressor object. Attributes in Python are different than methods and usually return a simple value or an array of values.

중요 참고: 이러한 계수를 얻기 위해 우리는 회귀자 객체에서 "coef_"와 "intercept_" 속성을 불렀다.<br> 파이썬의 속성은 메서드와 다르며 일반적으로 단순 값 또는 값 배열을 반환합니다.